pkg/client/cache/deltastore.go

```diff
@@ -0,0 +1,321 @@
+/*
+Copyright 2014 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/watch"
+)
+
+// DeltaStore is a Store implementation that provides a sequence of compressed events to a consumer
+// along with event types.  This differs from the FIFO implementation in that FIFO does not provide
+// events when an object is deleted and does not provide the type of event.  Events are compressed
+// in a manner similar to FIFO, but accounting for event types and deletions.  The exact
+// compression semantics are as follows:
+//
+// 1.  If a watch.Added is enqueued with state X and a watch.Modified with state Y is received,
+//     these are compressed into (Added, Y)
+//
+// 2.  If a watch.Added is enqueued with state X and a watch.Deleted is received, these are
+//     compressed and the item is removed from the queue
+//
+// 3.  If a watch.Modified is enqueued with state X and a watch.Modified with state Y is received,
+//     these two events are compressed into (Modified, Y)
+//
+// 4.  If a watch.Modified is enqueued with state X and a watch.Deleted is received, these are
+//     compressed into (Deleted, X)
```

Comment:
Assuming this is rebased onto #2294, then presumably the user either chose UID or Namespace+Name as the KeyFunc. 

If the user chose UID, then what you say should not happen (though it might if there was a restore of the apiserver state from a backup which caused things to go back in time;  crashing might be an okay response in that case though).  

If the user chose Namespace+Name, then they are saying that they don't want to do anything if the UID changes.  (Or they are confused.  Which I see as a bigger risk than poor performance.)